File: WarmUp/searchclient/searchclient/Box.java
package searchclient;

public class Box {
    public char chr;
    public Pair<Integer, Integer> location;
    public String color;

    Box(char chr, Pair<Integer, Integer> location, String color)
    {
        this.chr = chr;
        this.location = location;
        this.color = color;
    }

    Box(char chr, Pair<Integer, Integer> location)
    {
        this.chr = chr;
        this.location = location;
        this.color = "";
    }

    Box(Box box){
        this.chr = box.chr;
        this.location = box.location;
        this.color = "";
    }

    public void setLocation(Pair<Integer, Integer> location)
    {
        this.location = location;
    }

    @Override
    public int hashCode()
    {
        int result = 17;
        result = 31 * result + Character.getNumericValue(chr);
        result = 31 * result + location.getLeft();
        result = 31 * result + location.getRight();
        result = 31 * result + color.hashCode();
        return result;
    }

    @Override
    public boolean equals(Object o)
    {
        if (o == this) return true;
        if (!(o instanceof Box)) {
            return false;
        }

        Box box = (Box) o;

        return box.chr == chr && box.color.equals(color) && box.location.equals(location);
    } 
}

File: ProgrammingProject/searcher/src/Box.cpp

#include "Box.h"
#include <string>
#include <utility>

Box::Box(char chr, std::pair<int, int> location, std::string color)
{
    this->chr = chr;
    this->location = location;
    this->color = color;
}

Box::Box(char chr, std::pair<int, int> location)
{
    this->chr = chr;
    this->location = location;
    this->color = "";
}

Box::Box(Box * box){
    this->chr = box->chr;
    this->location = box->location;
    this->color = "";
}

void Box::setLocation(int x, int y)
{
  this->location = std::pair<int, int>(x,y);
}

std::pair<int, int> Box::getLocation(){
  return location;
}

int Box::hashCode()
{
    int result = 17;
    result = 31 * result + (int) (chr);
    result = 31 * result + std::get<0>(location);
    result = 31 * result + std::get<1>(location);
    result = 31 * result + std::hash<std::string>{}(color);
    return result;
}

bool Box::equals(Box * box)
{
    if (box == this) return true;
    return box->chr == chr && box->color == color && box->location == location;
}
int Box::getX(){
  return std::get<0>(location);
}
int Box::getY(){
  return std::get<1>(location);
}
